pkg/console/auth-config/auth-config.go

```diff
@@ -0,0 +1,203 @@
+func resyncPeriod(minResyncPeriod time.Duration) time.Duration {
+	factor := rand.Float64() + 1
+	return time.Duration(float64(minResyncPeriod.Nanoseconds()) * factor)
+}
+
+type reader struct {
+	informer cache.SharedIndexInformer
+	stop     chan struct{}
+	done     chan struct{}
+
+	lastRevision       string
+	userHeaders        []string
+	groupHeaders       []string
+	extraPrefixHeaders []string
+	clientCa           *x509.CertPool
+}
+
+var _ Reader = &reader{}
+
+func (r *reader) GetUserHeaders() ([]string, error) {
+	if err := r.refreshCache(); err != nil {
+		return nil, err
+	}
+	return r.userHeaders, nil
+}
+
+func (r *reader) GetGroupHeaders() ([]string, error) {
+	if err := r.refreshCache(); err != nil {
+		return nil, err
+	}
+	return r.groupHeaders, nil
+}
+
+func (r *reader) GetExtraHeaderPrefixes() ([]string, error) {
+	if err := r.refreshCache(); err != nil {
+		return nil, err
+	}
+	return r.extraPrefixHeaders, nil
+}
+
+func (r *reader) GetClientCA() (*x509.CertPool, error) {
+	if err := r.refreshCache(); err != nil {
+		return nil, err
+	}
+	return r.clientCa, nil
+}
+
+func (r *reader) Stop() {
+	close(r.stop)
+	<-r.done
+}
+
+func (r *reader) getConfigMap() (*v1.ConfigMap, error) {
+	configMap, exists, err := r.informer.GetStore().GetByKey(fmt.Sprintf("%s/%s", configMapNamespace, configMapName))
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("configmap %s/%s not found", configMapNamespace, configMapName)
+	}
+	return configMap.(*v1.ConfigMap), nil
+}
+
+func (r *reader) refreshCache() error {
+	configMap, err := r.getConfigMap()
+	if err != nil {
+		return err
+	}
+
+	if r.lastRevision == configMap.ResourceVersion {
+		return nil
+	}
+
+	userHeaders, err := getStringsFromConfigMap(configMap, usernameHeadersKey)
+	if err != nil {
+		return err
+	}
+
+	groupHeaders, err := getStringsFromConfigMap(configMap, groupHeadersKey)
+	if err != nil {
+		return err
+	}
+
+	extraPrefixHeaders, err := getStringsFromConfigMap(configMap, extraPrefixHeadersKey)
+	if err != nil {
+		return err
+	}
+
+	requestHeaderClientCA, ok := configMap.Data[clientCaKey]
+	if !ok {
+		return fmt.Errorf("requestheader-client-ca-file not found in configmap %s/%s", configMapNamespace, configMapName)
+	}
+
+	certs, err := cert.ParseCertsPEM([]byte(requestHeaderClientCA))
+	if err != nil {
+		return err
+	}
+	pool := x509.NewCertPool()
+	for _, crt := range certs {
+		pool.AddCert(crt)
+	}
+
+	r.userHeaders = userHeaders
+	r.groupHeaders = groupHeaders
+	r.extraPrefixHeaders = extraPrefixHeaders
+	r.clientCa = pool
+	r.lastRevision = configMap.ResourceVersion
+	return nil
+}
+
+func getStringsFromConfigMap(configMap *v1.ConfigMap, key string) ([]string, error) {
+	data, ok := configMap.Data[key]
+	if !ok {
+		return nil, fmt.Errorf("%s not found in configmap %s/%s", key, configMapNamespace, configMapName)
+	}
+	return unmarshalStrings(data)
+}
+
+func unmarshalStrings(strings string) ([]string, error) {
```

Comment:
This is only called by `getStringsFromConfigMap`, might make sense to combine these two funcs?

Reply:
Done